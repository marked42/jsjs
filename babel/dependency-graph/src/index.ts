import { ModuleSpecifier } from './Module'
import {
  ModuleResolver,
  ChainedModuleResolver,
  JavascriptModuleResolver,
} from './ModuleResolver'
import { DependencyGraphBuilder } from './DependencyGraph'

export function buildModuleDependencyGraph(entry: ModuleSpecifier) {
  const resolverChain: ModuleResolver[] = [new JavascriptModuleResolver()]
  const resolver = new ChainedModuleResolver(resolverChain)
  const builder = new DependencyGraphBuilder(resolver, entry)

  return builder.build()
}
