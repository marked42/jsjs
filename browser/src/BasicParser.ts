export class BasicParser {
  private index = 0
  constructor(private input: string) {}

  char() {
    return this.input[this.index]
  }

  next(len?: number) {
    len = len || 1
    this.index += len
  }

  eat(char?: string) {
    if (typeof char === 'string') {
      this.assert(char)
    }

    const text = char ? char : this.char()
    this.next(text ? text.length : 1)
    return text
  }

  assert(char: string) {
    if (!this.input.substring(this.index).startsWith(char)) {
      throw new Error(`unexpected char ${char}`)
    }
  }

  consumeWhitespace() {
    const whitespace = /[\t|\r|\n| ]/
    const startIndex = this.index
    while (whitespace.test(this.char())) {
      this.eat()
    }

    return this.input.substring(startIndex, this.index)
  }

  startsWith(input: string) {
    return this.input.substring(this.index).indexOf(input) === 0
  }

  end() {
    return this.index >= this.input.length
  }

  consumeCharsWhile(predicate: (char: string) => boolean) {
    let text = ''
    while (predicate(this.char()) && !this.end()) {
      text += this.eat()
    }

    return text
  }
}
