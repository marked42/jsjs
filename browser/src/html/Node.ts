export enum NodeType {
  Element,
  Text,
}

export interface TextNode {
  type: NodeType.Text
}

export abstract class Node {
  constructor(public type: NodeType) {}
}

export class TextNode extends Node {
  constructor(public textContent: string) {
    super(NodeType.Text)
  }
}

export class ElementNode extends Node {
  constructor(
    public name: string,
    public attributes: Record<string, string>,
    public children: Node[]
  ) {
    super(NodeType.Element)
  }
}
