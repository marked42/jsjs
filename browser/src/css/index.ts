import { BasicParser } from '../BasicParser'

class Stylesheet {
  constructor(public rules: Rule[]) {}
}

class Rule {
  constructor(
    public selectors: Selector[],
    public declarations: Declaration[]
  ) {}
}

class Selector {
  constructor(
    public tagName: string,
    public id: string,
    public classes: string[]
  ) {}
}

class Declaration {
  constructor(public name: string, public value: string) {}
}

enum DeclarationValueType {
  Keyword,
  Length,
  Color,
}

class DeclarationValue {
  constructor(public type: DeclarationValueType) {}
}

class DeclarationValueKeyword extends DeclarationValue {
  constructor(public value: string) {
    super(DeclarationValueType.Keyword)
  }
}

class DeclarationValueLength extends DeclarationValue {
  constructor(public value: number, public unit: Unit) {
    super(DeclarationValueType.Length)
  }
}

enum Unit {
  Px,
}

interface Color {
  r: number
  g: number
  b: number
  a: number
}

class DeclarationValueColor extends DeclarationValue {
  constructor(public value: Color) {
    super(DeclarationValueType.Color)
  }
}

export class CSSParser extends BasicParser {
  parseCSS() {}
}

export function parseCSS(input: string) {
  const parser = new CSSParser(input)
  return parser.parseCSS()
}
