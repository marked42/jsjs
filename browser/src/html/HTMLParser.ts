import { NodeType, Node, ElementNode, TextNode, NodeAttributeMap } from './Node'
import { BasicParser } from '../BasicParser'

function isHTMLElement(node: Node): node is ElementNode {
  return node.type === NodeType.Element && (node as ElementNode).name === 'html'
}

export default class HTMLParser extends BasicParser {
  parseHTML(): ElementNode {
    const nodes = this.parseNodes()

    if (nodes.length === 1) {
      const node = nodes[0]
      if (isHTMLElement(node)) {
        return node
      }
      throw new Error(`single root node must be html, get ${nodes[0]}`)
    }

    return new ElementNode('html', {}, nodes)
  }

  parseNodes(): Node[] {
    const nodes: Node[] = []

    for (;;) {
      if (this.startsWith('</') || this.end()) {
        break
      }
      nodes.push(this.parseNode())
    }

    return nodes
  }

  parseNode(): Node {
    if (this.startsWith('<')) {
      return this.parseElement()
    }

    return this.parseTextNode()
  }

  parseElement(): ElementNode {
    this.eat('<')
    const tagName = this.parseTagName()
    const attributes = this.parseTagAttributes()

    this.eat('>')

    const children = this.parseNodes()

    this.eat('<')
    this.eat('/')
    this.eat(tagName)
    this.eat('>')

    return new ElementNode(tagName, attributes, children)
  }

  parseTagName() {
    return this.consumeCharsWhile((char) => /[a-zA-Z0-9]/.test(char))
  }

  parseTagAttributes() {
    const attributes: NodeAttributeMap = {}
    for (;;) {
      if (this.char() === '>') {
        break
      }

      const { name, value } = this.parseTagAttribute()
      attributes[name] = value
    }
    return attributes
  }

  parseTagAttribute() {
    this.consumeWhitespace()
    const name = this.parseTagName()
    this.eat('=')
    const value = this.parseTagValue()
    return {
      name,
      value,
    }
  }

  parseTagValue() {
    const quote = this.char()
    if (quote !== '"' && quote !== "'") {
      throw new Error('expect quote')
    }
    this.next()

    let value = ''
    for (;;) {
      if (this.char() === quote) {
        this.next()
        break
      }
      value += this.char()
      this.next()
    }

    return value
  }

  parseTextNode() {
    const text = this.consumeCharsWhile((char) => char !== '<')

    return new TextNode(text)
  }
}

export function parseHTML(input: string) {
  const parser = new HTMLParser(input)

  return parser.parseHTML()
}
