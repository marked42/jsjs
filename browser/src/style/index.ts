import { ElementNode, Node, TextNode } from '../html/Node'
import { DeclarationValue, parseCSS, StyleSheet } from '../css'
import { parseHTML } from '../html'
import { MatchedRule } from '../css/MatchedRule'
import { Rule } from '../css/Rule'
import { Selector } from '../css/Selector'
import { Specificity } from '../css/Specificity'

interface PropertyMap {
  [key: string]: DeclarationValue
}

export class StyleNode {
  constructor(
    public node: Node,
    public specifiedValues: PropertyMap,
    public children: StyleNode[]
  ) {}

  lookup(name: string, name2: string, fallback: DeclarationValue) {
    return this.specifiedValues[name] || this.specifiedValues[name2] || fallback
  }

  value(name: string) {
    return this.specifiedValues[name]
  }
}

export function createStyleTree(html: string, css: string) {
  const node = parseHTML(html)
  const stylesheet = parseCSS(css)

  if (node instanceof TextNode) {
    return new StyleNode(node, {}, [])
  }

  return doCreateStyleTree(node, stylesheet)
}

// 类型推导问题 'doCreateStyleTree' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.ts(7023)
export function doCreateStyleTree(
  node: Node,
  stylesheet: StyleSheet
): StyleNode {
  if (node instanceof TextNode) {
    return new StyleNode(node, {}, [])
  }

  if (node instanceof ElementNode) {
    return new StyleNode(
      node,
      specifiedValuesForNode(node, stylesheet),
      node.children.map((child) => doCreateStyleTree(child, stylesheet))
    )
  }

  throw new Error(`unexpected node type ${node}`)
}

function specifiedValuesForNode(node: ElementNode, stylesheet: StyleSheet) {
  const matchedRules = matchRulesForNode(node, stylesheet)

  const specifiedValues: PropertyMap = {}
  for (const rule of matchedRules) {
    rule.rule.declarations.forEach((value) => {
      specifiedValues[value.name] = value.value
    })
  }

  return specifiedValues
}

function matchRulesForNode(
  node: ElementNode,
  stylesheet: StyleSheet
): MatchedRule[] {
  const matchedRules: MatchedRule[] = []

  for (const rule of stylesheet.rules) {
    const specificity = nodeMatchedWithRule(node, rule)
    if (specificity) {
      matchedRules.push(new MatchedRule(specificity, rule))
    }
  }

  matchedRules.sort((a, b) => a.specificity.compare(b.specificity))
  return matchedRules
}

function nodeMatchedWithRule(node: ElementNode, rule: Rule) {
  let specificity: Specificity | null = null

  for (const selector of rule.selectors) {
    if (selectorMatchedWithRule(node, selector)) {
      if (!specificity) {
        specificity = Specificity.from(selector)
      } else {
        const newSpecificity = Specificity.from(selector)
        if (specificity.compare(newSpecificity) < 0) {
          specificity = newSpecificity
        }
      }
    }
  }

  return specificity
}

function selectorMatchedWithRule(node: ElementNode, selector: Selector) {
  if (selector.tagName && selector.tagName !== node.name) {
    return false
  }

  if (selector.id && selector.id !== node.attributes['id']) {
    return false
  }

  const classAttributeValue = node.attributes['class']
  const classes: string[] =
    classAttributeValue === undefined ? [] : classAttributeValue.split(' ')
  if (
    selector.classes.length > 0 &&
    !classes.some((klass) => selector.classes.includes(klass))
  ) {
    return false
  }

  return true
}
