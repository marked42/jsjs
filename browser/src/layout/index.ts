import {
  DeclarationValue,
  DeclarationValueKeyword,
  DeclarationValueLength,
  Unit,
} from '../css'
import { StyleNode } from '../style'
import { Dimensions } from './Dimensions'

export enum LayoutBoxType {
  Block,
  Inline,
  Anonymous,
}

export class LayoutBox {
  public dimensions = Dimensions.default()
  public children = [] as LayoutBox[]
  constructor(public boxType: LayoutBoxType, public node?: StyleNode) {}

  getInlineContainer() {
    switch (this.boxType) {
      case LayoutBoxType.Inline:
      case LayoutBoxType.Anonymous:
        return this
      case LayoutBoxType.Block:
        const lastChild = this.children[this.children.length - 1]
        if (lastChild && lastChild.boxType === LayoutBoxType.Anonymous) {
          return lastChild
        } else {
          const anonymousBox = new LayoutBox(LayoutBoxType.Anonymous)

          this.children.push(anonymousBox)
          return anonymousBox
        }
    }
  }

  layout(containingBlock: Dimensions) {
    switch (this.boxType) {
      case LayoutBoxType.Block:
        this.layoutBlock(containingBlock)
      case LayoutBoxType.Inline:
      case LayoutBoxType.Anonymous:
        break
    }
  }

  layoutBlock(containingBlock: Dimensions) {
    this.calculateBlockWidth(containingBlock)
    this.calculateBlockPosition(containingBlock)
    this.layoutBlockChildren()
    this.calculateBlockHeight(containingBlock)
  }

  calculateBlockWidth(containingBlock: Dimensions) {
    const node = this.getStyleNode()

    const auto = DeclarationValueKeyword.of('auto')
    let width = node.value('width') || auto

    const zero = new DeclarationValueLength(0, Unit.Px)
    let marginLeft = node.lookup('margin-left', 'margin', zero)
    const borderLeft = node.lookup('border-left-width', 'border-width', zero)
    const paddingLeft = node.lookup('padding-left', 'padding', zero)
    let marginRight = node.lookup('margin-right', 'margin', zero)
    const borderRight = node.lookup('border-right-width', 'border-width', zero)
    const paddingRight = node.lookup('padding-right', 'padding', zero)

    const toPx = (value: DeclarationValue) => {
      if (value instanceof DeclarationValueLength) {
        return value.value
      }

      if (value instanceof DeclarationValueKeyword && value.value === 'auto') {
        return 0
      }

      throw new Error(`unexpected value ${value}`)
    }

    const totalWidth = [
      marginLeft,
      borderLeft,
      paddingLeft,
      width,
      paddingRight,
      borderRight,
      marginRight,
    ].reduce((acc, value) => {
      return acc + toPx(value)
    }, 0)

    const underflow = containingBlock.content.width - totalWidth

    // 把auto值转换为固定长度值
    // overflow
    if (underflow < 0) {
      if (marginLeft === auto) {
        marginLeft = new DeclarationValueLength(0, Unit.Px)
      }
      if (marginRight === auto) {
        marginRight = new DeclarationValueLength(0, Unit.Px)
      }
      if (width === auto) {
        width = new DeclarationValueLength(0, Unit.Px)
      } else {
        marginRight = new DeclarationValueLength(
          toPx(marginRight) + underflow,
          Unit.Px
        )
      }
    } else {
      // underflow 的8种情况
      if (width !== auto) {
        if (marginLeft === auto && marginRight === auto) {
          marginLeft = marginRight = new DeclarationValueLength(
            0.5 * underflow,
            Unit.Px
          )
        } else if (marginLeft === auto && marginRight !== auto) {
          marginLeft = new DeclarationValueLength(underflow, Unit.Px)
        } else if (marginLeft !== auto && marginRight === auto) {
          marginRight = new DeclarationValueLength(underflow, Unit.Px)
        } else if (marginLeft !== auto && marginRight !== auto) {
          marginRight = new DeclarationValueLength(
            toPx(marginRight) + underflow,
            Unit.Px
          )
        }
      } else {
        width = new DeclarationValueLength(underflow, Unit.Px)
        if (marginLeft === auto) {
          marginLeft = new DeclarationValueLength(0, Unit.Px)
        }
        if (marginRight === auto) {
          marginRight = new DeclarationValueLength(0, Unit.Px)
        }
      }
    }
  }

  getStyleNode() {
    if (this.boxType === LayoutBoxType.Anonymous || !this.node) {
      throw new Error('anonymous box has no style node')
    }

    return this.node
  }

  calculateBlockPosition(containingBlock: Dimensions) {}

  layoutBlockChildren() {
    this.children.forEach((child) => child.layoutBlock(this.dimensions))
  }

  calculateBlockHeight(containingBlock: Dimensions) {}
}

export function getLayoutBoxType(node: StyleNode) {
  const display = node.specifiedValues['display']

  if (display instanceof DeclarationValueKeyword) {
    if (display.value === 'block') {
      return LayoutBoxType.Block
    } else if (display.value === 'none') {
      return LayoutBoxType.Inline
    } else {
      return LayoutBoxType.Block
    }
  }
  return LayoutBoxType.Block
}

export function getDisplayValue(node: StyleNode) {
  const display = node.specifiedValues['display']

  if (display instanceof DeclarationValueKeyword) {
    if (display.value === 'block') {
      return 'block'
    } else if (display.value === 'none') {
      return 'none'
    } else {
      return 'inline'
    }
  }
  return 'inline'
}

export function buildLayoutTree(node: StyleNode) {
  const display = getDisplayValue(node)
  const root = new LayoutBox(LayoutBoxType.Block, node)
  switch (display) {
    case 'inline':
      root.boxType = LayoutBoxType.Inline
      break
    case 'block':
      root.boxType = LayoutBoxType.Block
      break
    case 'none':
      throw new Error('display none at root node not allowed')
  }

  node.children.forEach((child) => {
    const childDisplay = getDisplayValue(child)
    switch (childDisplay) {
      case 'block':
        root.children.push(buildLayoutTree(child))
        break
      case 'inline':
        root.getInlineContainer().children.push(buildLayoutTree(child))
      case 'none':
        break
    }
  })

  return root
}
