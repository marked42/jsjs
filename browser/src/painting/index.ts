import { LayoutBox, LayoutBoxType } from '../layout'
import { Rect } from '../layout/Rect'
import { Color, DeclarationValueColor } from '../css'

class DisplayCommand {}

class DisplayCommandSolidColor extends DisplayCommand {
  constructor(public color: Color, public rect: Rect) {
    super()
  }
}

export function buildDisplayList(layoutBox: LayoutBox): DisplayCommand[] {
  const displayList = [] as DisplayCommand[]
  renderLayoutBox(displayList, layoutBox)
  return displayList
}

export function renderLayoutBox(list: DisplayCommand[], layoutBox: LayoutBox) {
  renderBackground(list, layoutBox)
  renderBorders(list, layoutBox)

  layoutBox.children.forEach((child) => {
    renderLayoutBox(list, child)
  })
}

export function renderBackground(list: DisplayCommand[], layoutBox: LayoutBox) {
  const backgroundColor = getBoxColor(layoutBox, 'background')
  if (!backgroundColor) {
    return
  }

  list.push(
    new DisplayCommandSolidColor(
      backgroundColor,
      layoutBox.dimensions.borderBox()
    )
  )
}

export function renderBorders(list: DisplayCommand[], layoutBox: LayoutBox) {
  const borderColor = getBoxColor(layoutBox, 'border-color')
  if (!borderColor) {
    return
  }

  const dimensions = layoutBox.dimensions
  const borderBox = layoutBox.dimensions.borderBox()

  // left
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x,
        borderBox.y,
        dimensions.border.left,
        borderBox.height
      )
    )
  )

  // right
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x +
          dimensions.border.left +
          dimensions.padding.left +
          dimensions.content.width,
        borderBox.y,
        dimensions.border.right,
        borderBox.height
      )
    )
  )

  // top border
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(borderBox.x, borderBox.y, dimensions.border.top, borderBox.width)
    )
  )

  // bottom color
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x,
        borderBox.y + borderBox.height - dimensions.border.bottom,
        borderBox.width,
        dimensions.border.bottom
      )
    )
  )
}

export function getBoxColor(layoutBox: LayoutBox, name: string) {
  switch (layoutBox.boxType) {
    case LayoutBoxType.Block:
    case LayoutBoxType.Inline:
      const styleNode = layoutBox.getStyleNode()
      const value = styleNode.value(name)
      if (value) {
        if (value instanceof DeclarationValueColor) {
          return value.value
        } else {
          throw new Error(`style ${name} value is not color, get ${value}|`)
        }
      }
    case LayoutBoxType.Anonymous:
    default:
      break
  }
}

interface CanvasData {
  data: Uint8ClampedArray
  width: number
  height: number
}

export function drawLayoutBoxAsCanvasData(layoutBox: LayoutBox, bounds: Rect) {
  const canvasData = {
    data: new Uint8ClampedArray(4 * bounds.width * bounds.height),
    width: bounds.width,
    height: bounds.height,
  }

  const displayCommandList = buildDisplayList(layoutBox)

  for (const command of displayCommandList) {
    if (command instanceof DisplayCommandSolidColor) {
      applyCommandToCanvasData(command, canvasData)
    }
  }

  return canvasData
}

function applyCommandToCanvasData(
  command: DisplayCommandSolidColor,
  canvasData: CanvasData
) {
  for (let row = 0; row < canvasData.height; row++) {
    for (let col = 0; col < canvasData.width; col++) {
      canvasData.data[row * canvasData.width + col] = command.color.r
      canvasData.data[row * canvasData.width + col + 1] = command.color.g
      canvasData.data[row * canvasData.width + col + 2] = command.color.b
      canvasData.data[row * canvasData.width + col + 3] = command.color.a
    }
  }
}
