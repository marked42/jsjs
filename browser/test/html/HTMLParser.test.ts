import { parseHTML } from '../../src/html'

describe('parse html', () => {
  it('should parse html', () => {
    const input = `
			<div test="a">hello</div>
		`
    expect(parseHTML(input)).toMatchSnapshot()
  })
})
