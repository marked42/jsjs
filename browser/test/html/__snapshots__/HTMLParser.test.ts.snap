// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse html should parse html 1`] = `
ElementNode {
  "attributes": Object {},
  "children": Array [
    TextNode {
      "textContent": "
			",
      "type": 1,
    },
    ElementNode {
      "attributes": Object {
        "test": "a",
      },
      "children": Array [
        TextNode {
          "textContent": "hello",
          "type": 1,
        },
      ],
      "name": "div",
      "type": 0,
    },
    TextNode {
      "textContent": "
		",
      "type": 1,
    },
  ],
  "name": "html",
  "type": 0,
}
`;
